Test: Dynamically detect and set max-workers

sirreal opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Detect and set jest `maxWorkers` dynamically on CI.

## Why?

Jest workers are currently hard-coded at 2, which may not be optimal. [Jest suggests](https://jestjs.io/docs/troubleshooting#tests-are-extremely-slow-on-docker-andor-continuous-integration-ci-server) dynamically detecting and setting the worker count.

This seems to save roughly 50s on each JavaScript test run. There doesn't seem to be a significant different on mobile runs. It may be marginally slower.

---

I did 3 runs on this branch and compared with #59907 for a baseline. The unchanged tests have been (temporarily) removed from both branches to make results more accurate. You can compare all the timings [of this branch](https://github.com/WordPress/gutenberg/actions/runs/8296254381/usage) and the [baseline branch](https://github.com/WordPress/gutenberg/actions/runs/8296261073/usage).

The mobile tests seem mostly unchanged. However the JavaScript tests seem to run significantly faster on each run. Here are the runs and average for JavaScript (Node.js 20)

| Branch | Baseline |
|--------|--------|
| 4:07 | 4:47 |
| 4:02 | 4:53 |
| 3:53 | 4:52 |
| ~4:00 average | ~4:50 average |

Totals across three runs of Mobile + JavaScript (Node.js 20) + JavaScript (Node.js 21) a total savings of 4m33s.

| Branch | Baseline |
|--------|--------|
| 35:51 | 40:24 |
| ![Screenshot 2024-03-15 at 14 12 16](https://github.com/WordPress/gutenberg/assets/841763/5c8a8485-dec0-4f23-9032-9e6addafc025) | ![Screenshot 2024-03-15 at 14 11 29](https://github.com/WordPress/gutenberg/assets/841763/5852905f-54ed-4f15-8fc3-063c1a18ee85) |

sirreal committed:
Use SimenB/github-actions-cpu-cores@v2 to dynamically detect and set maxWorkers

See:
https://jestjs.io/docs/troubleshooting#tests-are-extremely-slow-on-docker-andor-continuous-integration-ci-server

sirreal committed:
DROPME: Remove non-unit-tests for testing

sirreal committed:
DROPME: remove non-js tests

sirreal committed:
Revert "DROPME: remove non-js tests"

This reverts commit e81545ffb602a644bb7df7ce147829a871718b13.

sirreal committed:
Revert "DROPME: Remove non-unit-tests for testing"

This reverts commit df103e37e0b73c8c53df6d4a6ceba28adb7c1844.

sirreal commented:
I'm not sure how many workers we have access to, we use a lot, but we could consider sharding the tests as well.

desrosj commented on .github/workflows/unit-test.yml:
```diff
@@ -39,17 +39,21 @@ jobs:
                   node-version: ${{ matrix.node }}
 
+            - name: Get number of CPU cores
+              uses: SimenB/github-actions-cpu-cores@v2
```
```suggestion
              uses: SimenB/github-actions-cpu-cores@97ba232459a8e02ff6121db9362b09661c875ab8 # v2.0.0
```
For third party actions, we should hardcode the SHA value for the version we wish to use. This is recommended in the [security hardening guidelines](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions). We use Dependabot to manage these updates automatically, so the burden of maintenance is not that great.

desrosj commented on .github/workflows/unit-test.yml:
```diff
@@ -293,6 +297,10 @@ jobs:
                   show-progress: ${{ runner.debug == '1' && 'true' || 'false' }}
 
+            - name: Get number of CPU cores
+              uses: SimenB/github-actions-cpu-cores@v2
```
```suggestion
              uses: SimenB/github-actions-cpu-cores@97ba232459a8e02ff6121db9362b09661c875ab8 # v2.0.0
```

desrosj reviewed (changes requested):
A few small changes, and I think this is worth merging to try out.

As for the sharding recommendation and splitting into additional jobs, I do recommend we do that as well.

As an organization, we have an allotment of 500 concurrent runners. At peak times, it's not uncommon to have jobs waiting for runner machines. I haven't done the math in at least 6 months, but my findings previously were that the majority of our workflow jobs complete in 1-2 minutes or less. So while there is an initial spike in runner allocation when a commit is made or a PR is updated, it's usually fairly brief.

Personally, the first thing I would do is split off the date tests into it's own job. While it's only 5 seconds, it makes sense to split it off since it's a separate group entirely. When they fail, it makes it more obvious which group has issues.

Looking at your observations for baseline and the results of this PR, I wouldn't be surprised if we could get this down under 2 minutes (or even 1) per runner if we shard into 2, 3, or 4 different runners. This does require more runner machines, but if they take less than 1 minute, it shouldn't result in any meaningful delay.

sirreal committed:
Pin github-actions-cpu-cores action to sha

Pin to

https://github.com/SimenB/github-actions-cpu-cores/commit/97ba232459a8e02ff6121db9362b09661c875ab8

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

sirreal commented:
Thanks for the review and thoughts, @desrosj! I've applied your suggestion to pin the action to the specific sha.

If you agree, I'd like to merge this PR as-is and iteratively work on other improvements.

sirreal commented:
I'll continue with some improvements discussed in #60045.
